File: nexus-sdk/go/nexus/transaction.go
package nexus

import (
	"fmt"
	"math/big"
)

// TransactionManager builds transactions against real nexus-core envelopes
// and queries confirmed transactions.
type TransactionManager struct {
	client *Client
}

// Transaction is an unsigned NEX transfer built from real chain state.
type Transaction struct {
	From     string
	To       string
	Value    *big.Int
	Token    string
	Nonce    int64
	GasPrice *big.Int
}

// BuildTransfer assembles an unsigned transfer: nonce comes from
// nexus_getTransactionCount ({"count":N} envelope), gasPrice from the
// wallet-service/oracle recommendation (core JSON-RPC has no nexus_gasPrice;
// getNodeStatus carries no gasPrice field — see JsonRpcController.doGetNodeStatus).
func (tm *TransactionManager) BuildTransfer(from, to string, amount *big.Int) (*Transaction, error) {
	if !ValidateAddress(from) {
		return nil, fmt.Errorf("nexus: invalid from address %q", from)
	}
	if !ValidateAddress(to) {
		return nil, fmt.Errorf("nexus: invalid to address %q", to)
	}
	nonce, err := tm.client.Wallet.GetNonce(from)
	if err != nil {
		return nil, err
	}
	return &Transaction{
		From:     from,
		To:       to,
		Value:    amount,
		Token:    "NEX",
		Nonce:    nonce,
		GasPrice: big.NewInt(0), // gas 费表见 core Transaction.GAS_TABLE；SDK 侧无查询接口
	}, nil
}

// GetTransactionByHash returns a confirmed transaction
// (nexus_getTransactionByHash → {txHash, from, to, amount, status, ...}).
// NOT a receipt: pending/confirmation semantics don't apply to core's RPC
// (it returns only on-chain transactions with status "success").
func (tm *TransactionManager) GetTransactionByHash(txHash string) (map[string]interface{}, error) {
	result, err := tm.client.RPCCall("nexus_getTransactionByHash", txHash)
	if err != nil {
		return nil, fmt.Errorf("nexus: get transaction: %w", err)
	}
	if m, ok := result.(map[string]interface{}); ok {
		return m, nil
	}
	return nil, fmt.Errorf("nexus: unexpected transaction envelope: %T", result)
}

// GetLatestTransactions returns recent on-chain transactions
// (nexus_getLatestTransactions, limit clamped to 1..100 server-side).
func (tm *TransactionManager) GetLatestTransactions(limit int) ([]map[string]interface{}, error) {
	result, err := tm.client.RPCCall("nexus_getLatestTransactions", limit)
	if err != nil {
		return nil, fmt.Errorf("nexus: latest transactions: %w", err)
	}
	arr, ok := result.([]interface{})
	if !ok {
		return nil, fmt.Errorf("nexus: unexpected list envelope: %T", result)
	}
	out := make([]map[string]interface{}, 0, len(arr))
	for _, e := range arr {
		if m, ok := e.(map[string]interface{}); ok {
			out = append(out, m)
		}
	}
	return out, nil
}

// BridgeManager queries the on-chain cross-chain bridge view.
type BridgeManager struct {
	client *Client
}

// CrossChainTransaction mirrors a toRpcCrossChainTx entry (BRIDGE_* payload
// projection over the last 200 blocks).
type CrossChainTransaction struct {
	BridgeTxID   string `json:"bridgeTxId"`
	SourceChain  string `json:"sourceChain"`
	TargetChain  string `json:"targetChain"`
	Recipient    string `json:"recipient"`
	Amount       string `json:"amount"`
	Status       string `json:"status"`
	TimeStamp    int64  `json:"timestamp"`
}

// List returns cross-chain transactions (nexus_getCrossChainTransactions).
// statusFilter is optional ("pending"/"confirmed"/... — server-side filter).
func (b *BridgeManager) List(limit int, statusFilter string) ([]map[string]interface{}, error) {
	var result interface{}
	var err error
	if statusFilter == "" {
		result, err = b.client.RPCCall("nexus_getCrossChainTransactions", limit)
	} else {
		result, err = b.client.RPCCall("nexus_getCrossChainTransactions", limit, statusFilter)
	}
	if err != nil {
		return nil, fmt.Errorf("nexus: cross-chain list: %w", err)
	}
	arr, ok := result.([]interface{})
	if !ok {
		return nil, fmt.Errorf("nexus: unexpected cross-chain envelope: %T", result)
	}
	out := make([]map[string]interface{}, 0, len(arr))
	for _, e := range arr {
		if m, ok := e.(map[string]interface{}); ok {
			out = append(out, m)
		}
	}
	return out, nil
}

File: nexus-sdk/python/nexus_sdk/transaction.py
"""NexusChain SDK 交易模块（真实信封契约）。

构建转账（nonce 取自 nexus_getTransactionCount 真实信封）、查询确认交易。
签名/广播不是本模块职责——走 wallet-service（见 WalletManager.submit_transfer）。
"""

from typing import Any, Dict, List, Optional

from .address import validate_address


class TransferRequest:
    """未签名的 NEX 转账（字段来自真实链上查询）。"""

    def __init__(self, from_addr: str, to: str, amount: int, nonce: int):
        self.from_addr = from_addr
        self.to = to
        self.amount = amount
        self.nonce = nonce

    def __repr__(self) -> str:  # pragma: no cover — debug 辅助
        return (f"TransferRequest(from={self.from_addr!r}, to={self.to!r}, "
                f"amount={self.amount}, nonce={self.nonce})")


class TransactionManager:
    """交易构建与查询。"""

    def __init__(self, client):
        self._client = client

    def build_transfer(self, from_addr: str, to: str, amount: int) -> TransferRequest:
        """构建转账：nonce 实时取链（{"count":N} 信封）。

        本地先校验地址格式（Base58 + 25 字节 + keccak 校验尾）。
        """
        if not validate_address(from_addr):
            raise ValueError(f"invalid from address: {from_addr!r}")
        if not validate_address(to):
            raise ValueError(f"invalid to address: {to!r}")
        nonce = self._client.wallet.get_nonce(from_addr)
        return TransferRequest(from_addr, to, amount, nonce)

    def get_transaction_by_hash(self, tx_hash: str) -> Optional[Dict[str, Any]]:
        """查确认交易（nexus_getTransactionByHash）。

        注意这不是回执：core RPC 只返回已上链交易（status 恒 "success"）。
        """
        result = self._client.rpc_call("nexus_getTransactionByHash", [tx_hash])
        if isinstance(result, dict):
            return result
        return None

    def get_latest_transactions(self, limit: int = 20) -> List[Dict[str, Any]]:
        """最新交易列表（nexus_getLatestTransactions，服务端夹逼 1..100）。"""
        result = self._client.rpc_call("nexus_getLatestTransactions", [limit])
        if not isinstance(result, list):
            raise ValueError(f"unexpected list envelope: {type(result)}")
        return [tx for tx in result if isinstance(tx, dict)]
